protocols/noise/src/keys.rs

```diff
@@ -18,156 +18,185 @@
+    /// Get a reference to the static public key.
+    pub fn public(&self) -> &StaticPublicKey {
+        &self.public
     }
-}
 
-impl<T> AsRef<[u8]> for PublicKey<T> {
-    fn as_ref(&self) -> &[u8] {
-        &self.bytes
+    /// Create a new static X25519 keypair.
+    pub fn new() -> StaticKeypair {
+        Self::from(x25519::StaticSecret::new(&mut rand::thread_rng()))
     }
-}
 
-/// ECC secret key.
-#[derive(Clone)]
-pub struct SecretKey {
-    scalar: Scalar
+    /// Construct a static X25519 keypair from an Ed25519 keypair.
+    ///
+    /// *Note*: If the Ed25519 keypair is already used in the context
+    /// of other cryptographic protocols outside of Noise, e.g. for
+    /// signing in the `secio` protocol, it should be preferred to
+    /// create a new `StaticKeypair` for use in the Noise protocol.
+    ///
+    /// See also:
+    ///
+    /// [Noise: Static Key Reuse](http://www.noiseprotocol.org/noise.html#security-considerations)
+    /// [Ed25519 to Curve25519](https://libsodium.gitbook.io/doc/advanced/ed25519-curve25519)
+    pub fn from_ed25519(ed: &ed25519::Keypair) -> StaticKeypair {
+        // An Ed25519 public key is derived off the left half of the SHA512 of the
+        // secret scalar, hence a matching conversion of the secret key must do
+        // the same to yield a Curve25519 keypair with the same public key.
+        let ed25519_sk = ed.secret_bytes();
+        let mut curve25519_sk: [u8; 32] = [0; 32];
+        let hash = digest(&SHA512, ed25519_sk);
+        curve25519_sk.copy_from_slice(&hash.as_ref()[..32]);
+        Self::from(x25519::StaticSecret::from(curve25519_sk))
+    }
 }
 
-impl SecretKey {
-    pub(crate) fn new(bytes: [u8; 32]) -> Self {
-        SecretKey {
-            scalar: Scalar::from_bytes_mod_order(bytes)
-        }
+impl From<x25519::StaticSecret> for StaticKeypair {
+    fn from(secret: x25519::StaticSecret) -> StaticKeypair {
+        let public = StaticPublicKey(x25519::PublicKey::from(&secret));
+        StaticKeypair { secret, public }
     }
 }
 
-impl SecretKey {
-    /// Compute the public key from this secret key.
-    ///
-    /// Performs scalar multiplication with Curve25519's base point.
-    pub fn public(&self) -> PublicKey<Curve25519> {
-        PublicKey::new(self.x25519(&X25519_BASEPOINT).0)
-    }
+/// Static X25519 public key.
+pub struct StaticPublicKey(pub x25519::PublicKey);
 
-    /// Elliptic-Curve Diffie-Hellman key agreement.
+impl StaticPublicKey {
+    /// Construct a static X25519 public key from an Ed25519 public key.
+    ///
+    /// See also [`StaticKeypair::from_ed25519`].
     ///
-    /// Performs scalar multiplication with the given public key.
-    pub fn ecdh(&self, pk: &PublicKey<Curve25519>) -> [u8; 32] {
-        self.x25519(&MontgomeryPoint(pk.bytes)).0
+    /// [`StaticKeypair::from_ed25519`]: struct.StaticKeypair.html#method.from_ed25519
+    pub fn from_ed25519(pk: &ed25519::PublicKey) -> Option<StaticPublicKey> {
```

Comment:
An ed25519 deserialization can fail when the point lies on the twist.  Actually x25519 too but x25519 does not consider this an error.

In this case, `ed25519::PublicKey` seemingly wraps an `ed25519_dalek::PublicKey`, which previously wrapped just a `[u8; 32]`.  In that state, there is no reason to believe decompression occurred previously, so this could error if it's the first decompression.  

I think `ed25519_dalek::PublicKey` now does decompression earlier, but not sure the version being used here, or if they might change their mind again in future.

Reply:
Yes, I had assumed that the Edwards key would be uncompressed because it is `encode`d first and then decompressed again and the conversion to Montgomery should be infallible if done directly. But I have not looked into `core::identity::ed25519` yet.

Reply:
Indeed [PublicKey::from_bytes](https://docs.rs/ed25519-dalek/1.0.0-pre.1/src/ed25519_dalek/public.rs.html#128-144) already does the decompression, and that is also used in [Keypair::from_bytes](https://docs.rs/ed25519-dalek/1.0.0-pre.1/src/ed25519_dalek/ed25519.rs.html#200-211), so decompression should actually be infallible at this point.

Reply:
In general, you cannot necessarily rely on that always being the case because some libraries might do this differently.